src/encode.cpp

```diff
@@ -970,6 +970,55 @@ void encodeOp(State &st, mlir::tosa::BitwiseXorOp op, bool) {
       [](auto &&a, auto &&b) { return (Expr)a ^ (Expr)b; });
 }
 
+template<>
+void encodeOp(State &st, mlir::tosa::Conv2DOp op, bool) {
+  auto input = st.regs.get<Tensor>(op.input());
+  auto weight = st.regs.get<Tensor>(op.weight());
+  auto bias = st.regs.get<Tensor>(op.bias());
+  vector<Expr> strides, pad, dilations;
+
+  for (auto s: op.stride()) {
+    strides.push_back(Index(s.dyn_cast<mlir::IntegerAttr>().getInt()));
+  };
+
+  for (auto s: op.dilation()) {
+    dilations.push_back(Index(s.dyn_cast<mlir::IntegerAttr>().getInt()));
+  };
+
+  for (auto s: op.pad()) {
+    pad.push_back(Index(s.dyn_cast<mlir::IntegerAttr>().getInt()));
+  };
+
+  assert(strides.size() == 2 && dilations.size() == 2 && pad.size() == 4);
+  auto elemTy = getElemTy(op.getResult());
+
+  // input rank should be 4
+  vector<Expr> padInd = Index::boundIndexVars(input.getRank());
+  vector<Expr> srcDims = input.getDims();
+
+  vector<Expr> srcInd = {padInd[0], padInd[1] - pad[0],
+                            padInd[2] - pad[2], padInd[3]};
+
+  vector<Expr> padDims = {srcDims[0], srcDims[1] + pad[0] + pad[1],
+                            srcDims[2] + pad[2] + pad[3], srcDims[3]};
+
+  auto cond = padInd[1].uge(pad[0]) & padInd[1].ult(pad[0] + srcDims[1]) &
+                padInd[2].uge(pad[2]) & padInd[2].ult(pad[2] + srcDims[2]);
+
+  Expr output = Expr::mkIte(cond, input.get(srcInd).first, *getZero(elemTy))
+                + bias.get({padInd[3]}).first;
+  Expr init = Expr::mkIte(cond, Expr::mkBool(true), input.isInitialized(padInd));
+
+  auto padInput = Tensor::mkLambda(
```

Comment:
You can use `Tensor::mkInitializedLambda` here

Reply:
Also, please add `st.wellDefined(op, {input, weight, bias}.isFullyInitialized())`